File: src/core/currency/currency.go
package currency

import (
	"errors"
	"strconv"
	"strings"
)

// CurrencyInfo holds information about supported currencies
type CurrencyInfo struct {
	Symbol          string
	Name            string
	SmallUnitName   string
	SmallUnitFactor int64
	SupportedChains []string
}

// SupportedCurrencies maps currency symbols to their info
var SupportedCurrencies = map[string]CurrencyInfo{
	"ETH": {
		Symbol:          "ETH",
		Name:            "Ethereum",
		SmallUnitName:   "wei",
		SmallUnitFactor: 1000000000000000000, // 1e18
		SupportedChains: []string{"base", "ethereum"},
	},
	// Stubbed currencies for forward compatibility
	"BTC": {
		Symbol:          "BTC",
		Name:            "Bitcoin",
		SmallUnitName:   "sats",
		SmallUnitFactor: 100000000, // 1e8
		SupportedChains: []string{"bitcoin"},
	},
	"ALGO": {
		Symbol:          "ALGO",
		Name:            "Algorand",
		SmallUnitName:   "microalgos",
		SmallUnitFactor: 1000000, // 1e6
		SupportedChains: []string{"algorand"},
	},
	"SOL": {
		Symbol:          "SOL",
		Name:            "Solana",
		SmallUnitName:   "lamports",
		SmallUnitFactor: 1000000000, // 1e9
		SupportedChains: []string{"solana"},
	},
}

// Currency represents a currency amount with both major and minor units
type Currency struct {
	Symbol     string
	Amount     string // Major unit (e.g., "1.5" ETH)
	SmallUnit  string // Minor unit (e.g., "1500000000000000000" wei)
	Blockchain string
}

// NewCurrency creates a new Currency instance
func NewCurrency(symbol, amount, smallUnit, blockchain string) (*Currency, error) {
	info, exists := SupportedCurrencies[strings.ToUpper(symbol)]
	if !exists {
		return nil, errors.New("unsupported currency: " + symbol)
	}

	// Validate blockchain support (only for implemented currencies)
	if symbol == "ETH" {
		validChain := false
		for _, chain := range info.SupportedChains {
			if strings.ToLower(blockchain) == chain {
				validChain = true
				break
			}
		}
		if !validChain {
			return nil, errors.New("currency " + symbol + " not supported on blockchain " + blockchain)
		}
	}

	return &Currency{
		Symbol:     strings.ToUpper(symbol),
		Amount:     amount,
		SmallUnit:  smallUnit,
		Blockchain: strings.ToLower(blockchain),
	}, nil
}

// ConvertToSmallUnit converts major unit to small unit
func (c *Currency) ConvertToSmallUnit() (string, error) {
	if c.SmallUnit != "" {
		return c.SmallUnit, nil
	}

	info, exists := SupportedCurrencies[c.Symbol]
	if !exists {
		return "", errors.New("unsupported currency: " + c.Symbol)
	}

	amount, err := strconv.ParseFloat(c.Amount, 64)
	if err != nil {
		return "", errors.New("invalid amount: " + c.Amount)
	}

	smallUnitAmount := int64(amount * float64(info.SmallUnitFactor))
	return strconv.FormatInt(smallUnitAmount, 10), nil
}

// ConvertToMajorUnit converts small unit to major unit
func (c *Currency) ConvertToMajorUnit() (string, error) {
	if c.Amount != "" {
		return c.Amount, nil
	}

	info, exists := SupportedCurrencies[c.Symbol]
	if !exists {
		return "", errors.New("unsupported currency: " + c.Symbol)
	}

	smallUnitAmount, err := strconv.ParseInt(c.SmallUnit, 10, 64)
	if err != nil {
		return "", errors.New("invalid small unit amount: " + c.SmallUnit)
	}

	majorAmount := float64(smallUnitAmount) / float64(info.SmallUnitFactor)
	return strconv.FormatFloat(majorAmount, 'f', -1, 64), nil
}

// GetSmallUnitName returns the name of the small unit (e.g., "wei", "sats")
func (c *Currency) GetSmallUnitName() string {
	info, exists := SupportedCurrencies[c.Symbol]
	if !exists {
		return "units"
	}
	return info.SmallUnitName
}

// IsSupported checks if a currency is supported
func IsSupported(symbol string) bool {
	_, exists := SupportedCurrencies[strings.ToUpper(symbol)]
	return exists
}

// IsImplemented checks if a currency is fully implemented (not just stubbed)
func IsImplemented(symbol string) bool {
	return strings.ToUpper(symbol) == "ETH"
}

// ValidateForBlockchain checks if currency is supported on the given blockchain
func ValidateForBlockchain(symbol, blockchain string) error {
	info, exists := SupportedCurrencies[strings.ToUpper(symbol)]
	if !exists {
		return errors.New("unsupported currency: " + symbol)
	}

	// Only validate for implemented currencies
	if IsImplemented(symbol) {
		for _, chain := range info.SupportedChains {
			if strings.ToLower(blockchain) == chain {
				return nil
			}
		}
		return errors.New("currency " + symbol + " not supported on blockchain " + blockchain)
	}

	return nil // Allow stubbed currencies for forward compatibility
}

// FormatDisplay returns a human-readable format of the currency
func (c *Currency) FormatDisplay() string {
	if c.Amount != "" {
		return c.Amount + " " + c.Symbol
	}

	majorUnit, err := c.ConvertToMajorUnit()
	if err != nil {
		return c.SmallUnit + " " + c.GetSmallUnitName()
	}

	return majorUnit + " " + c.Symbol
}

File: src/typescript/util/currency.ts
import {LogError} from "./log";

// Currency information for supported currencies
interface CurrencyInfo {
    symbol: string;
    name: string;
    smallUnitName: string;
    smallUnitFactor: number;
    supportedChains: string[];
}

// Supported currencies configuration
const SUPPORTED_CURRENCIES: Record<string, CurrencyInfo> = {
    ETH: {
        symbol: "ETH",
        name: "Ethereum", 
        smallUnitName: "wei",
        smallUnitFactor: 1e18,
        supportedChains: ["base", "ethereum"]
    },
    // Stubbed currencies for forward compatibility
    BTC: {
        symbol: "BTC",
        name: "Bitcoin",
        smallUnitName: "sats", 
        smallUnitFactor: 1e8,
        supportedChains: ["bitcoin"]
    },
    ALGO: {
        symbol: "ALGO",
        name: "Algorand",
        smallUnitName: "microalgos",
        smallUnitFactor: 1e6,
        supportedChains: ["algorand"]
    },
    SOL: {
        symbol: "SOL", 
        name: "Solana",
        smallUnitName: "lamports",
        smallUnitFactor: 1e9,
        supportedChains: ["solana"]
    }
};

// Currency class for handling multi-currency amounts
export class Currency {
    public symbol: string;
    public amount: string;      // Major unit (e.g., "1.5" ETH)
    public smallUnit: string;   // Minor unit (e.g., "1500000000000000000" wei)
    public blockchain: string;

    constructor(symbol: string, amount: string, smallUnit: string, blockchain: string) {
        this.symbol = symbol.toUpperCase();
        this.amount = amount;
        this.smallUnit = smallUnit;
        this.blockchain = blockchain.toLowerCase();
    }

    // Convert major unit to small unit
    public convertToSmallUnit(): string {
        if (this.smallUnit) {
            return this.smallUnit;
        }

        const info = SUPPORTED_CURRENCIES[this.symbol];
        if (!info) {
            LogError("Unsupported currency: " + this.symbol);
            return "";
        }

        const amount = parseFloat(this.amount);
        if (isNaN(amount)) {
            LogError("Invalid amount: " + this.amount);
            return "";
        }

        const smallUnitAmount = Math.floor(amount * info.smallUnitFactor);
        return smallUnitAmount.toString();
    }

    // Convert small unit to major unit
    public convertToMajorUnit(): string {
        if (this.amount) {
            return this.amount;
        }

        const info = SUPPORTED_CURRENCIES[this.symbol];
        if (!info) {
            LogError("Unsupported currency: " + this.symbol);
            return "";
        }

        const smallUnitAmount = parseInt(this.smallUnit, 10);
        if (isNaN(smallUnitAmount)) {
            LogError("Invalid small unit amount: " + this.smallUnit);
            return "";
        }

        const majorAmount = smallUnitAmount / info.smallUnitFactor;
        return majorAmount.toString();
    }

    // Get the name of the small unit
    public getSmallUnitName(): string {
        const info = SUPPORTED_CURRENCIES[this.symbol];
        return info ? info.smallUnitName : "units";
    }

    // Format for display
    public formatDisplay(): string {
        if (this.amount) {
            return `${this.amount} ${this.symbol}`;
        }

        const majorUnit = this.convertToMajorUnit();
        if (majorUnit) {
            return `${majorUnit} ${this.symbol}`;
        }

        return `${this.smallUnit} ${this.getSmallUnitName()}`;
    }
}

// Utility functions
export function isSupported(symbol: string): boolean {
    return SUPPORTED_CURRENCIES[symbol.toUpperCase()] !== undefined;
}

export function isImplemented(symbol: string): boolean {
    return symbol.toUpperCase() === "ETH";
}

export function validateForBlockchain(symbol: string, blockchain: string): boolean {
    const info = SUPPORTED_CURRENCIES[symbol.toUpperCase()];
    if (!info) {
        return false;
    }

    // Only validate for implemented currencies
    if (isImplemented(symbol)) {
        return info.supportedChains.includes(blockchain.toLowerCase());
    }

    return true; // Allow stubbed currencies for forward compatibility
}

export function getSupportedCurrencies(): string[] {
    return Object.keys(SUPPORTED_CURRENCIES);
}

export function getImplementedCurrencies(): string[] {
    return ["ETH"];
}

export function getCurrencyInfo(symbol: string): CurrencyInfo | null {
    return SUPPORTED_CURRENCIES[symbol.toUpperCase()] || null;
}

// Convert wei to ETH (for backward compatibility)
export function weiToEth(wei: string): string {
    const currency = new Currency("ETH", "", wei, "base");
    return currency.convertToMajorUnit();
}

// Convert ETH to wei (for backward compatibility)
export function ethToWei(eth: string): string {
    const currency = new Currency("ETH", eth, "", "base");
    return currency.convertToSmallUnit();
}